components/brave_news/browser/feed_fetcher.cc

```diff
@@ -152,47 +156,72 @@ void FeedFetcher::OnFetchFeedFetchedFeed(
     return;
   }
 
-  std::move(callback).Run({locale, etag, ParseFeedItems(result.value_body())});
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce([](base::Value value) { return ParseFeedItems(value); },
+                     std::move(result.value_body())),
+      base::BindOnce(
+          [](std::string locale, std::string etag,
+             FetchFeedSourceCallback callback,
+             std::vector<mojom::FeedItemPtr> items) {
+            std::move(callback).Run(
+                {std::move(locale), std::move(etag), std::move(items)});
+          },
+          std::move(locale), std::move(etag), std::move(callback)));
 }
 
 void FeedFetcher::OnFetchFeedFetchedAll(FetchFeedCallback callback,
                                         Publishers publishers,
                                         std::vector<FeedSourceResult> results) {
-  std::size_t total_size = 0;
-  for (const auto& result : results) {
-    total_size += result.items.size();
-  }
-  VLOG(1) << "All feed item fetches done with item count: " << total_size;
-
-  ETags etags;
-  FeedItems feed;
-  feed.reserve(total_size);
-
-  // We want to deduplicate the feed, as the feeds for different regions **may**
-  // have overlap.
-  base::flat_set<GURL> seen;
-
-  for (auto& result : results) {
-    etags[result.key] = result.etag;
-    for (auto& item : result.items) {
-      GURL url;
-      if (item->is_article()) {
-        url = item->get_article()->data->url;
-      } else if (item->is_promoted_article()) {
-        url = item->get_promoted_article()->data->url;
-      }
-
-      // Skip this, we've already seen it.
-      if (!url.is_empty() && seen.contains(url)) {
-        continue;
-      }
-      seen.insert(url);
-
-      feed.push_back(std::move(item));
-    }
-  }
-
-  std::move(callback).Run(std::move(feed), std::move(etags));
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(
+          [](std::vector<FeedSourceResult> results) {
+            std::size_t total_size = 0;
+            for (const auto& result : results) {
+              total_size += result.items.size();
+            }
+            VLOG(1) << "All feed item fetches done with item count: "
+                    << total_size;
+
+            ETags etags;
+            FeedItems feed;
+            feed.reserve(total_size);
+
+            // We want to deduplicate the feed, as the feeds for different
+            // regions
+            // **may** have overlap.
+            base::flat_set<GURL> seen;
```

Comment:
I think this was the main factor in the performance issues - when I was looking at a trace most of the time was spent comparing `GURL`s

Reply:
the operator==() overload for GURL is using string::operator==. So I guess there wouldn't be big difference. https://source.chromium.org/chromium/chromium/src/+/44812c2055259e0e48f5077280d918ab64ab390b:url/gurl.cc;l=544

Reply:
I was thinking that maybe it does `.spec()` for every comparison, so in `flat_set` the worst case would be `O(n)` calls to `spec()`

Maybe using `std::set` would be better here?

Reply:
Yeah, flat_set is designed to be used for handful data. [As per the document](https://source.chromium.org/chromium/chromium/src/+/3e815ef723d2b1a4b467fff93e16a177ebae1626:base/containers/README.md;l=77), it's O(n) on insertion and deletion. 

And std::set is implemented as RB tree, so it's performance would be O(logn) in the worst case. so std::unordered_set, which has O(1) complexity, could be even better

Reply:
Done! I'm doing `.reserve(n)` now too - as I understand it'll make `std::unordered_set` reserve enough space to fit `n` items at the `load_factor` of the set.

Reply:
Wow, didn't know about it. Nice catch! It seems that reserving would help us to avoid rehashing everything 👍

Reply:
for 3000+ GURL items `unordered_set` looks like a best choice.
BTW, it's not clear for me why the data URLs used as a key here (not something like id or hash).

Reply:
hm, url_hash also a super long string ;(
<img width="995" alt="image" src="https://github.com/brave/brave-core/assets/45488748/06d59f33-10f8-45d3-a162-3b1a7c2160f4">

Reply:
So the reason I wasn't using `url_hash` is that in the past the backend has been very inconsistent in providing it so it hasn't been safe to rely on.

Unfortunately, that makes url the closest thing to a guaranteed Id we have.